### ISSUE TITLE
Can't retract facts.

### ISSUE THREAD

**Issue by d4hines** [tau-prolog/tau-prolog#208]
Hi!

Thanks for making Tau Prolog! It's an awesome system! I'm prototyping a new action language with it over at https://flamingo-lang.org.

I'm having trouble using `retract/1`. For example, in this program: http://tau-prolog.org/sandbox/7sS9oHzN, I would expect the query `foo.` to return `false`, but it returns `true`.

When I try using `retract` or `asserta`  I get a permission error. E.g., the query:
```
asserta(bar(3)). bar(X).
```
produces the following answers:
```
X = 2;
uncaught exception: error(permission_error(modify, static_procedure, bar/1), /(asserta, 1))
```

What am I doing wrong?
Thanks!

**Comment by d4hines** [comment]
Actually, I think I figured it out:

```
:-dynamic(bar/1).
```

It might be worth adding this to the documentation.

**Comment by pmoura** [comment]
It's actually a (standards compliance) bug. Asserting a (previously unknown) predicate implicitly makes it dynamic. That said, declaring all dynamic predicates using the `dynamic/1` directive is good practice.

**Comment by jariazavalverde** [comment]
The execution is as expected. You must explicitly declare dynamic predicates if you provide any clauses for them. Also, in your example, note that the only way to execute a goal from the program is through the `initialization` directive.

```prolog
:- dynamic(bar/1).
:- initialization(retract(bar(_))).

bar(2).
foo(X) :- bar(X).
```

The `initialization` directive will be available from the next version v0.3.0 (in the current version of git it is already available).

**Comment by pmoura** [comment]
@jariazavalverde is correct. I didn't look into the program. As he explained, if a file contains clauses for a dynamic predicate, it must also contain a `dynamic/1` directive for the predicate as per the ISO Prolog standard. Also, your program is trying to redefine the `rectract/1` built-in predicate.
